# Post Insights Daily

##### In postinsightsdaily and videoinsightsdaily we retreive 45 days of data for each video
###### Like in postinsights daily we get 45 days views for video (from one of the columns)
###### Say video was published on 21th june 2020 so our database should have 45 days of data for that scripts that is  till 4th aug data should be there.
###### Our task is to check if there is the data o all 45 days or not

In [1]:
import pandas as pd
import mysql.connector

In [2]:
from collections import Counter

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
# database connection
cnx = mysql.connector.connect(host = 'mysqlservernewjprod.mysql.database.azure.com', user = "phantom@mysqlservernewjprod", password = 'Zurich$1', db = 'fb', port = 3306)

In [5]:
cursor = cnx.cursor(buffered=True)

In [6]:
sql = 'SELECT * FROM fb.postinsightsdaily where pageName = "NEWJPLUS";' #sql query for retreiving the data

df = pd.read_sql(sql, cnx) # retreive the data

df.to_csv('postinsightsdaily.csv', index=False) # save the retreived data locally

cursor.close() 

True

In [7]:
data = pd.read_csv('postinsightsdaily.csv') # read the loacally saved data
data.head(3)

id          videoId            pageId  \
0  07cdf894-5c53-11ea-9de5-48682eb0e778  121907679248283  1954072204870360   
1  07ce8f39-5c53-11ea-9de5-48682eb0e778  121907679248283  1954072204870360   
2  07cf387b-5c53-11ea-9de5-48682eb0e778  121907679248283  1954072204870360   

   pageName                                      permalink_url  \
0  newjplus  https://www.facebook.com/NEWJplus/videos/12190...   
1  newjplus  https://www.facebook.com/NEWJplus/videos/12190...   
2  newjplus  https://www.facebook.com/NEWJplus/videos/12190...   

                                             message    post_created_time  \
0  पीएम मोदी वाराणसी से विकास परियोजनाओं की शुरुआ...  2020-02-16 06:51:07   
1  पीएम मोदी वाराणसी से विकास परियोजनाओं की शुरुआ...  2020-02-16 06:51:07   
2  पीएम मोदी वाराणसी से विकास परियोजनाओं की शुरुआ...  2020-02-16 06:51:07   

  post_created_time_zone   published_time_ist  post_video_views  \
0                 +00:00  2020-02-16 12:21:07                 1   
1                 +00:00  2020-02-16 12:21:07              1638   
2                 +00:00  2020-02-16 12:21:07              5775   

  post_video_views_end_time post_video_views_end_time_ist  \
0       2020-02-15 08:00:00           2020-02-15 13:30:00   
1       2020-02-16 08:00:00           2020-02-16 13:30:00   
2       2020-02-17 08:00:00           2020-02-17 13:30:00   

   post_video_view_time post_video_view_time_end_time  \
0                177869           2020-02-15 08:00:00   
1              42348390           2020-02-16 08:00:00   
2             202808193           2020-02-17 08:00:00   

  post_video_view_time_end_time_ist  post_video_view_time_organic  \
0               2020-02-15 13:30:00                        177869   
1               2020-02-16 13:30:00                      42348390   
2               2020-02-17 13:30:00                     202808193   

  post_video_view_time_organic_end_time  \
0                   2020-02-15 08:00:00   
1                   2020-02-16 08:00:00   
2                   2020-02-17 08:00:00   

  post_video_view_time_organic_end_time_ist  post_video_views_10s_paid  \
0                       2020-02-15 13:30:00                          0   
1                       2020-02-16 13:30:00                          0   
2                       2020-02-17 13:30:00                          0   

  post_video_views_10s_paid_end_time post_video_views_10s_paid_end_time_ist  \
0                2020-02-15 08:00:00                    2020-02-15 13:30:00   
1                2020-02-16 08:00:00                    2020-02-16 13:30:00   
2                2020-02-17 08:00:00                    2020-02-17 13:30:00   

   post_video_views_paid post_video_views_paid_end_time  \
0                      0            2020-02-15 08:00:00   
1                      0            2020-02-16 08:00:00   
2                      0            2020-02-17 08:00:00   

  post_video_views_paid_end_time_ist  post_video_views_organic  \
0                2020-02-15 13:30:00                         1   
1                2020-02-16 13:30:00                      1638   
2                2020-02-17 13:30:00                      5775   

  post_video_views_organic_end_time post_video_views_organic_end_time_ist  \
0               2020-02-15 08:00:00                   2020-02-15 13:30:00   
1               2020-02-16 08:00:00                   2020-02-16 13:30:00   
2               2020-02-17 08:00:00                   2020-02-17 13:30:00   

   post_video_views_10s post_video_views_10s_end_time  \
0                     1           2020-02-15 08:00:00   
1                   680           2020-02-16 08:00:00   
2                  2263           2020-02-17 08:00:00   

  post_video_views_10s_end_time_ist consolidated_end_time  \
0               2020-02-15 13:30:00            2020-02-15   
1               2020-02-16 13:30:00            2020-02-16   
2               2020-02-17 13:30:00            2020-02-17   

                                               token      

In [10]:
data.columns

Index(['id', 'videoId', 'pageId', 'pageName', 'permalink_url', 'message',
       'post_created_time', 'post_created_time_zone', 'published_time_ist',
       'post_video_views', 'post_video_views_end_time',
       'post_video_views_end_time_ist', 'post_video_view_time',
       'post_video_view_time_end_time', 'post_video_view_time_end_time_ist',
       'post_video_view_time_organic', 'post_video_view_time_organic_end_time',
       'post_video_view_time_organic_end_time_ist',
       'post_video_views_10s_paid', 'post_video_views_10s_paid_end_time',
       'post_video_views_10s_paid_end_time_ist', 'post_video_views_paid',
       'post_video_views_paid_end_time', 'post_video_views_paid_end_time_ist',
       'post_video_views_organic', 'post_video_views_organic_end_time',
       'post_video_views_organic_end_time_ist', 'post_video_views_10s',
       'post_video_views_10s_end_time', 'post_video_views_10s_end_time_ist',
       'consolidated_end_time', 'token', 'updated_at', 'created_at'],
   

In [11]:
data.shape

(242180, 34)

### Task 1: check the total number of unique video Id's

In [12]:
# check unique video id's
# the number of unique video id's means number of videos because each video id represent one unique video
data['videoId'].nunique()

5611

`Total 5611 videos data retreived`

### Task 2: check the last date for which data is retreived

In [13]:
# check the data type of 'video published time'
data['published_time_ist'].dtype 

dtype('O')

In [14]:
# covert the data['published_time_ist'] to datetime format
data['published_time_ist'] = pd.to_datetime(data['published_time_ist'])

In [15]:
# now check the last date for which data is retreived
data['published_time_ist'].max()

Timestamp('2020-10-13 21:00:00')

`Last date for which data was populated 13th oct 2020 which is (today-2)`

In [16]:
data['published_time_ist'].min() # Initial date for which we have data in database

Timestamp('2019-02-01 09:00:00')

###### Any video published before  data['published_time_ist'].max() - 45 days should have the 45 days data

### Task 3: Check if the videos published before 45 days have 45 observations or not

In [19]:
from datetime import datetime, timedelta

In [20]:
# check the date, before which any video published should have 45 days record
data['published_time_ist'].max() - timedelta(days = 45)

Timestamp('2020-08-29 21:00:00')

In [21]:
# create a dataframe of the videos for which data should be populated 
df_must_45_days = data[data['published_time_ist'] < data['published_time_ist'].max() - timedelta(days = 45)]

In [22]:
df_must_45_days['published_time_ist'].max() # max date

Timestamp('2020-08-29 19:05:00')

In [23]:
Counter(df_must_45_days['videoId'])

Counter({121907679248283: 45,
         131251801583746: 45,
         159309575071366: 45,
         169926994296431: 45,
         171037124192609: 45,
         177365873532288: 45,
         180124883239041: 45,
         181491583127053: 46,
         185784969170417: 45,
         188988899165178: 29,
         195832038433525: 46,
         197138025018068: 45,
         204377750880700: 45,
         228401808304999: 45,
         230281634927035: 45,
         234976301256198: 45,
         240165207239097: 45,
         245043266485381: 45,
         246147033338652: 43,
         267802947484871: 45,
         274134603629494: 45,
         276159787034729: 29,
         283732025942811: 45,
         306119667422427: 45,
         322281705593049: 45,
         329656738404054: 45,
         336853384169775: 45,
         338251397054002: 45,
         347532586061866: 45,
         352198196176133: 45,
         361098111232348: 45,
         373218390199958: 45,
         380020009542335: 45,
         4

In [34]:
index_id_less_than_45 = [i for i in range(len(list(Counter(df_must_45_days['videoId']).values()))) if list(Counter(df_must_45_days['videoId']).values())[i]<45]

In [35]:
ids_with_less_than_45 = [list(Counter(df_must_45_days['videoId']).keys())[i] for i in index_id_less_than_45]

In [37]:
len(ids_with_less_than_45)

185

`There are total 185 videos for which data isn't retreived for 45 days`

In [38]:
# check the list of the Id's for which data is not populated correctly
ids_with_less_than_45

[188988899165178,
 246147033338652,
 276159787034729,
 726264157916141,
 729109411190690,
 740372620057665,
 747734206000187,
 793020047918937,
 954315911693792,
 1192429324434437,
 1626054707567665,
 2459086174368958,
 2459951354282440,
 2465292237081685,
 2469208170023425,
 2479520118992230,
 2480601745550734,
 2637219943222246,
 2637300913214149,
 2637339306543643,
 2637408176536756,
 2637479939862913,
 2637995443144696,
 2638068189804088,
 2638173149793592,
 2638185716459002,
 2638244729786434,
 2638247226452851,
 2638282469782660,
 2638353259775581,
 2638868676390706,
 2639054839705423,
 2639197796357794,
 2639210999689807,
 2639237133020527,
 2639282789682628,
 2639825946294979,
 2639951012949139,
 2639992572944983,
 2640038009607106,
 2640172206260353,
 2640181039592803,
 2640185306259043,
 2640881016189472,
 2640926982851542,
 2640961896181384,
 2641016746175899,
 2674479769496263,
 2674531149491125,
 2674541859490054,
 2674600132817560,
 2674611349483105,
 2674623152815258,
 2

In [41]:
# check one of these Id's for verification
data[data['videoId'] == 188988899165178]

id          videoId            pageId  \
406  05faffec-b2d4-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
407  05fc2956-b2d4-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
408  05fcb1e5-b2d4-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
409  d66861db-b498-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
410  d668ec82-b498-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
411  a3cf72af-b62a-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
412  a3cffa32-b62a-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
413  e9e1d952-b7bc-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
414  e9e257f6-b7bc-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
415  f4b2965b-b94e-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
416  f4b309df-b94e-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
417  a0e08738-baaf-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
418  a0e10c7b-baaf-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
419  b6a45c83-bc41-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
420  b6a4dd39-bc41-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
421  d7c095a2-bdd3-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
422  d7c11e59-bdd3-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
423  070e295a-bf66-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
424  070eb371-bf66-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
425  3a870160-c0f8-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
426  3a878406-c0f8-11ea-b179-ecc92973c78a  188988899165178  1954072204870360   
427  6f1c17fe-c28a-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
428  6f1c9b18-c28a-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
429  99667bf7-c41c-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
430  99670676-c41c-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
431  ca870363-c5ae-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
432  ca878bab-c5ae-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
433  f7a6c259-c740-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   
434  f7a7440a-c740-11ea-a721-c08f8b0c98d5  188988899165178  1954072204870360   

     pageName                                      permalink_url  \
406  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
407  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
408  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
409  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
410  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
411  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
412  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
413  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
414  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
415  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
416  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
417  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
418  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
419  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
420  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
421  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
422  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
423  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
424  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
425  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
426  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
427  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
428  newjplus  https://www.facebook.com/NEWJplus/videos/18898...   
429  n

In [42]:
data[data['videoId'] == 188988899165178].shape

(29, 34)

`This video has been updated for 29 days only`

In [44]:
# last day of video got updated
data[data['videoId'] == 188988899165178]['post_video_views_end_time_ist'].max()

'2020-07-15 12:30:00'

In [45]:
# check the date till when video should be updated
data[data['videoId'] == 188988899165178]['published_time_ist'].min() + timedelta(days = 45)

Timestamp('2020-08-01 15:25:28')

`Video should have been updated till 2020-08-01 but its last updated date is 2020-07-15`

### Task 4: Check if there is any missing data

In [46]:
data.isnull().sum()

id                                           0
videoId                                      0
pageId                                       0
pageName                                     0
permalink_url                                0
message                                      0
post_created_time                            0
post_created_time_zone                       0
published_time_ist                           0
post_video_views                             0
post_video_views_end_time                    0
post_video_views_end_time_ist                0
post_video_view_time                         0
post_video_view_time_end_time                0
post_video_view_time_end_time_ist            0
post_video_view_time_organic                 0
post_video_view_time_organic_end_time        0
post_video_view_time_organic_end_time_ist    0
post_video_views_10s_paid                    0
post_video_views_10s_paid_end_time           0
post_video_views_10s_paid_end_time_ist       0
post_video_vi

`no data is missing`